sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -146,21 +159,119 @@ public String getMimeType() {
    */
   protected final WritableByteChannelFactory writableByteChannelFactory;
 
+
   /**
-   * Base filename for final output files.
+   * A naming policy for output files.
    */
-  protected final ValueProvider<String> baseOutputFilename;
+  public abstract static class FilenamePolicy
+      implements org.apache.beam.sdk.transforms.SerializableFunction<
+       FilenamePolicy.Context, String> {
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * Window and pane will only be provided if windowed writes have been requested using
+     * {@link Write.Bound#withWindowedWrites}. The policy must produce unique filenames for unique
+     * Context objects.
+     */
+    public static class Context {
+      @Nullable private BoundedWindow window;
+      @Nullable private PaneInfo paneInfo;
+      private int shardNumber;
+      private int numShards;
+
+      public Context(
```

Comment:
Can you explain what you're suggesting? How does making parameters non-nullable help? 

In truth, the user can always shoot themselves in the foot with the policy (e.g. generating the same filename for different panes of the same window). Making parameters non-nullable does not really fix this. IMO we simply need to document very well what is required here.